File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
def solution(maps):
    answer = 0
    bfs(maps)
    if maps[len(maps)-1][len(maps[0])-1] == 1:
        return -1
    return maps[len(maps)-1][len(maps[0])-1]


def bfs(maps):
    que = deque()
    que.append((0,0))
    dx = [-1,1,0,0]
    dy=[0,0,-1,1]
    n = len(maps)
    m = len(maps[0])
    visit = [[False]*m for _ in range(n)]
    visit[0][0]=True
    while que:
        x,y = que.popleft()
        for i in range(4):
            nx=x+dx[i]
            ny=y+dy[i]
            if 0<=nx<m and 0<=ny<n and maps[ny][nx] ==1:
                if not visit[ny][nx]:
                    maps[ny][nx] = maps[y][x]+1
                    visit[ny][nx] =True
                    que.append((nx,ny))
    

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;

class Solution {
    
    static int[] dx = {0,-1,0,1};
    static int[] dy = {-1,0,1,0};
    static boolean[][] visit;
    
    static int n;
    static int m;
    
    static class Node{
        int i;
        int j;
        int dist;
        Node(int i, int j, int dist){
            this.i = i;
            this.j = j;
            this.dist = dist;
        }
        
        
        
    }
    
    
    public int solution(int[][] maps) {
        int answer = 0;
        
        n = maps.length;
        m = maps[0].length;
        
        visit = new boolean [n][m];
        answer = bfs(maps);
        
        return answer;
    }
    
    public int bfs(int[][] maps){
        
        visit[0][0] = true;
        Queue<Node> que = new LinkedList();
        
        que.add(new Node(0,0,1));
        
        while(!que.isEmpty()){
            Node tmp = que.poll();
            int i = tmp.i;
            int j = tmp.j;
            int dist = tmp.dist;
            
            if(i == n-1 && j == m-1){
                return dist;
            }
            
            for(int d=0;d<4;d++){
                int nx = j+dx[d];
                int ny = i+dy[d];
                
                if(nx<0 || nx>=m || ny<0 || ny>=n) continue;
                
                if(!visit[ny][nx] && maps[ny][nx] == 1){
                    visit[ny][nx] = true;
                    
                    que.add(new Node(ny,nx,dist+1));
                }
                    
            }
            
        }
        
        return -1;
        
    }
    
    
}